modxcms/revolution
# Add new error reporting for grids

### What does it do?
Changed to base grids class `onStoreException` method to provide in-grid reporting of errors.

### Why is it needed?
Picking up where @Mark-H started, this displays an error message where previously a grid was mysteriously empty when a server error occurred. Note, however, this solution shows stack traces in the browser console by default and only outputs them to screen when MODx is in debug mode. Otherwise, an elegant message is shown in the grid, notifying users of an error and suggesting ways to view more detailed information.

Here are some screenshots of various scenarios when this solution is applied:

![modx-pr-example-outputs--xl](https://user-images.githubusercontent.com/689075/207939464-3860b90f-2fdc-422e-9253-f164c03c1cee.jpg)

### How to test
1. Clear all caches and run a rebuild of the default template.
2. Break a GetList processor. I used the Contexts grid for testing and added this line before the return in the `prepareRow` method: `$contextArray = $this->noExist($contextArray);`
3. Test the output in your test grid by changing your php `display_errors` and MODx `debug` system setting to various values. Note that stack traces will only appear when `display_errors` is set to "on" and `debug` is not "0". Verify receiving messages as shown in the example image above.
4. Note that I've found that the various caches (modx, browser, etc) can be finicky and it may take multiple clears when changing your config scenarios to get the expected output.

### Related issue(s)/PR(s)
Alternate solution to the withdrawn PR #16293.
